### PR TITLE
[GeoMechanicsApplication] Add 1D functionality to the existing 2D/3D thermal element

### LINKED ISSUE #12259: [GeoMechanicsApplication] Add 1D functionality to the existing 2D/3D thermal element
### Description

The objective of this issue is to design a well-element, which is inherently a line element. It has large similarities with the thermal line element, except the constutative law and the way of handeling with geometrical parameters such as Jacobian. So, it can be considered as an extention in capabilities of the existing thermal element. 

In oder to avoid duplication in the code, we decided to used the existing thermal element, and extend its capability to handle with the line elementas. To make it more generic and resonable, the thermal element needs to be extended to be able to solve the standard case in 1D. Then the well-element can be added on the top of it.

### PR DESCRIPTION
**Pull request by mnabideltares**
**📝 Description**
The objective of this issue is to design a well-element, which is inherently a line element. It has large similarities with the thermal line element, except the constutative law and the way of handeling with geometrical parameters such as Jacobian. So, it can be considered as an extention in capabilities of the existing thermal element.

In oder to avoid duplication in the code, we decided to used the existing thermal element, and extend its capability to handle with the line elementas. To make it more generic and resonable, the thermal element needs to be extended to be able to solve the standard case in 1D. Then the well-element can be added on the top of it.

**🆕 Changelog**
- Registered thermal line element
- Modified the existing thermal element to handle thermal line element
- Added test cases

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h [hunk 1]
```diff
@@ -236,9 +259,10 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) TransientThermalElement : public Ele
         const Vector&                                    rIntegrationCoefficients,
         const ProcessInfo&                               rCurrentProcessInfo) const
     {
-        GeoThermalDispersionLaw law{TDim};
-        const auto              constitutive_matrix =
-            law.CalculateThermalDispersionMatrix(GetProperties(), rCurrentProcessInfo);
+        std::size_t NumberOfDimensions = GetGeometry().LocalSpaceDimension();
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h]: Can be const and since it's a local variable, it should be snake case
```suggestion
        const std::size_t number_of_dimensions = GetGeometry().LocalSpaceDimension();
```

File: applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h [hunk 2]
```diff
@@ -86,6 +94,22 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) TransientThermalElement : public Ele
...
+                return GeometryData::IntegrationMethod::GI_GAUSS_2;
+            case 4:
+                return GeometryData::IntegrationMethod::GI_GAUSS_3;
+            case 5:
+                return GeometryData::IntegrationMethod::GI_GAUSS_5;
+            default:
+                KRATOS_ERROR << "Can't return integration method: unexpected number of nodes: " << TNumNodes
+                             << std::endl;
+            }
+        }
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:111]: This switch statement is the same for the thermal element and the 1d pw element. I would suggest to extract a utility function for this for the PR that goes back latest. I would suggest to try and first merge the thermal element and afterwards the Pw, since this one is way smaller and more straight-forward

File: applications/GeoMechanicsApplication/geo_mechanics_application.cpp [hunk 3]
```diff
@@ -220,6 +220,13 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_ELEMENT("GeoTransientThermalElement3D20N", mTransientThermalElement3D20N)
     KRATOS_REGISTER_ELEMENT("GeoTransientThermalElement3D27N", mTransientThermalElement3D27N)
 
+    KRATOS_REGISTER_ELEMENT("GeoTransientThermalLineElement2D2N", mTransientThermalLineElement2D2N)
+    KRATOS_REGISTER_ELEMENT("GeoTransientThermalLineElement2D3N", mTransientThermalLineElement2D3N)
+    KRATOS_REGISTER_ELEMENT("GeoTransientThermalLineElement2D4N", mTransientThermalLineElement2D4N)
+    KRATOS_REGISTER_ELEMENT("GeoTransientThermalLineElement2D5N", mTransientThermalLineElement2D5N)
+    KRATOS_REGISTER_ELEMENT("GeoTransientThermalLineElement3D2N", mTransientThermalLineElement3D2N)
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/geo_mechanics_application.cpp:227]: Is there a reason 2D goes until 5N and 3D only to 3N?

File: applications/GeoMechanicsApplication/tests/test_transient_thermal.py
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/test_transient_thermal.py:1]: Could you add a readme.md for these tests, or add it to the existing readme of the thermal element tests?

**Review by rfaasse — COMMENTED** [review]
This is a very nicely scoped change, which makes it easy to review. I have some minor suggestions and questions, but I would suggest to merge this one first, before the Pw element

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp [hunk 4]
```diff
@@ -49,16 +49,19 @@ Matrix GeoThermalDispersionLaw::CalculateThermalDispersionMatrix(const Propertie
...
+
+    if (mNumberOfDimensions >= 2) {
+        const auto y  = static_cast<int>(indexThermalFlux::Y);
+        result(x, y) = solid_fraction * rProp[THERMAL_CONDUCTIVITY_SOLID_XY];
+        result(y, y) = solid_fraction * rProp[THERMAL_CONDUCTIVITY_SOLID_YY] +
+                       water_fraction * rProp[THERMAL_CONDUCTIVITY_WATER];
+        result(y, x) = result(x, y);
+    }
 
     if (mNumberOfDimensions == 3) {
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp]: If this is placed inside the previous if, there is no extra retest and the redefinition of y is not necessary.

File: applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h [hunk 5]
```diff
@@ -70,8 +70,16 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) TransientThermalElement : public Ele
 
         GeometryType::ShapeFunctionsGradientsType dN_dX_container;
         Vector                                    det_J_container;
-        GetGeometry().ShapeFunctionsIntegrationPointsGradients(dN_dX_container, det_J_container,
-                                                               GetIntegrationMethod());
+
+        if (GetGeometry().LocalSpaceDimension() == 1) {
+            GetGeometry().DeterminantOfJacobian(det_J_container, this->GetIntegrationMethod());
+            dN_dX_container = GetGeometry().ShapeFunctionsLocalGradients(this->GetIntegrationMethod());
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:76]: Do these 2 lines also work for LocalSpacDimension 2 and 3? That would make the if else construction redundant.

File: applications/GeoMechanicsApplication/geo_mechanics_application.cpp (hunk 3 above)
Comment (mnabideltares) [reply applications/GeoMechanicsApplication/geo_mechanics_application.cpp:227]: @rfaasse  Thank you for the nice review.

We don't have geometries Line3D4N and Line3D5N. Because we don't have a such high order 3D tetrahedrons or hexahedrons.

**Review by WPK4FEM — COMMENTED** [review]
Dear Mohammed,
Thank you for this short and effective implementation. Apart from the remarks I made, please add the extra tests you made in some way to the README you made for the other thermal elements.
Wijtze Pieter

File: applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp [hunk 6]
```diff
@@ -17,13 +17,17 @@
 namespace Kratos
 {
 
+template class TransientThermalElement<2, 2>;
 template class TransientThermalElement<2, 3>;
 template class TransientThermalElement<2, 4>;
+template class TransientThermalElement<2, 5>;
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:23]: Just to check my understanding: the only reason we don't have a 5-noded thermal line element in **3D** is because there are no other 3D elements with 5 nodes along each side. Is that correct?

File: applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h (hunk 1 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h]: I think you may have forgotten to change the variable name to snake case...?

File: applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_2D2N/test_thermal_line_element_2D2N.mdpa [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    4
+  End SubModelPartNodes
+  Begin SubModelPartElements
+    1
+    2
+    3
+  End SubModelPartElements
+  Begin SubModelPartConditions
+  End SubModelPartConditions
+End SubModelPart
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_2D2N/test_thermal_line_element_2D2N.mdpa]: Sub-model part `Gravity` doesn't appear to be used at all. Could you please verify whether that is correct? If so, please remove it.

File: applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_3D2N/MaterialParameters.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
+{
+  "properties": [{
+    "model_part_name"                      : "PorousDomain.filter",
+    "properties_id"                        : 1,
+    "Material" : {
+      "constitutive_law": {
+        "name"                             : "GeoThermalDispersion2DLaw"
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_3D2N/MaterialParameters.json]: For consistency, I would prefer to use `"GeoThermalDispersion3DLaw"` here (this is a 3D test, after all).

File: applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_3D2N/ProjectParameters.json [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    "start_time"                         : 0.0,
+    "end_time"                           : 86400000,
+    "echo_level"                         : 0,
+    "parallel_type"                      : "OpenMP",
+    "number_of_threads"                  : 1
+  },                                     
+  "solver_settings": {                   
+    "solver_type"                        : "T",
+    "model_part_name"                    : "PorousDomain",
+    "domain_size"                        : 2,
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_3D2N/ProjectParameters.json]: Shouldn't this value be 3?

File: applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_3D2N/test_thermal_line_element_3D2N.mdpa [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    4
+  End SubModelPartNodes
+  Begin SubModelPartElements
+    1
+    2
+    3
+  End SubModelPartElements
+  Begin SubModelPartConditions
+  End SubModelPartConditions
+End SubModelPart
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_3D2N/test_thermal_line_element_3D2N.mdpa]: This sub-model part doesn't appear to be used. If that is correct, please remove it.

File: applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_3D3N/MaterialParameters.json [hunk 11]
```diff
@@ -0,0 +1,30 @@
+{
+  "properties": [{
+    "model_part_name"                      : "PorousDomain.filter",
+    "properties_id"                        : 1,
+    "Material" : {
+      "constitutive_law": {
+        "name"                             : "GeoThermalDispersion2DLaw"
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_3D3N/MaterialParameters.json]: Again, I would prefer to use the `"GeoThermalDispersion3DLaw"` here.

File: applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_3D3N/ProjectParameters.json [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    "start_time"                         : 0.0,
+    "end_time"                           : 86400000,
+    "echo_level"                         : 0,
+    "parallel_type"                      : "OpenMP",
+    "number_of_threads"                  : 1
+  },                                     
+  "solver_settings": {                   
+    "solver_type"                        : "T",
+    "model_part_name"                    : "PorousDomain",
+    "domain_size"                        : 2,
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/test_thermal_line_element_3D3N/ProjectParameters.json]: Shouldn't this value be equal to 3?

**Review by avdg81 — COMMENTED** [review]
A very nice and clean extension of your previous work. Well done! I only have a few minor comments.

File: applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp (hunk 6 above)
Comment (mnabideltares) [reply applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:23]: @avdg81  thank you for the review. 
It is right. We don't have 3D elements with edges of 5 nodes, so we didn't build a such line elements. It is mainly because:
1. We usually used line elements (or more accurately, geometry) for conditions. The condition needs to match with the element order. As we haven't 3D element with a such high order, we never built 3D high order line element.
2. In the case of beams or sheet piles, they need to be connected to soil element. And again they need to match in order.
However, for purly 1D element like thge one in this issue, it is possible to extend and use 3D4N , or 3D5N. It is maybe an issue for future (probably with low priority, as it is not a bloker)

File: applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h (hunk 5 above)
Comment (mnabideltares) [reply applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:76]: @WPK4FEM thank you for the review.
I tried to use it also for 2D and 3D, but it seems "dN_dX_container" gets wrong values.

File: applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/README.md [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+## Case Specification
+In this thermal test case, a 3 m deep soil is considered, with everywhere set to 0 degrees then a sudden jump at the top and bottom boundary temperatures are given. These are -10 degrees at bottom and 100 degrees at the top boundaries. The simulation spans 1000 days to allow for a transition from an exponential to a linear temperature profile between the two sides. This test is conducted for various configurations, including 2D2N, 2D3N, 2D4N, 2D5N, 3D2N and 3D3N line elements. The temperature distribution along the depth is then evaluated with its own result.
+The boundary conditions are shown below:
+
+<img src="../documentation_data/test_thermal_line_element.svg" alt="Visualization of the Boundary conditions" title="Visualization of the Boundary conditions" width="600">
+
+## Results
+
+The picture below illustrates the temperature contours resulting from the simulation (as an example the 2D3N test is shown below).
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/README.md:15]: In my opinion it would be helpful to add a sentence to the results section that explains that the assert you are doing in the test itself (checking the $-10 + (100 -- 10)/3 = 26.66667$ on the second node).

**Review by rfaasse — COMMENTED** [review]
Thank you for processing the review comments and adding the readme for the tests, this turned out a very comprehensive PR! 

I have one minor comment to add some clarification for what the test actually asserts, next to the illustrative graph you added, but other than that I think we're good to go!

File: applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/README.md (hunk 13 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_line_element/README.md:15]: In addition to Richard's comment, I believe it would be helpful to explicitly state in the text that the displayed temperature contour plot was created for the final time step.

**Review by avdg81 — COMMENTED** [review]
Thank you so much for processing the review comments. Your extension is clear, clean, well-tested and well-documented. I only have one minor (non-blocking) comment.

**Review by rfaasse — APPROVED** [review]
Looks good to me!

**Review by WPK4FEM — APPROVED** [review]
For me this looks ready to merge.
